#01. Importing libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

#02. Importing data 

In [3]:
#create string path for main project folder
path = r'/Users/AngieUS/Desktop/Instacart Project'

#import orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3_orders_products', 'orders.csv'), index_col = False)

#check df_ords
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [4]:
#import products data
df_prods = pd. read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3_orders_products', 'products.csv'), index_col = False)

#check df_prods
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

#03. Data wrangling procedures

In [26]:
#dropping columns - remove eval_set column from df_ords visually
df_ords.drop(columns = ['eval_set'])

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_last_order  
0                          NaN  
1                         15.0  
2                         21.0  
3                         29.0  
4                         28.0  
...                        ...  
3421078                   29.0  
3421079                   30.0  
3421080                   18.0  
3421081                    7.0  
3421082                   30.0  

[3421083 rows x 6 columns]

In [27]:
#dropping columns - overwrite dataset to actually remove the column
df_ords = df_ords.drop(columns = ['eval_set'])

In [6]:
#finding missing values - NaN
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [7]:
#renaming columns
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

#check df_ords columns
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [8]:
#changing a variable's data type
df_ords['order_id'] = df_ords['order_id'].astype('str')

#check df_ords 'order_id' data type
df_ords['order_id'].dtype

dtype('O')

In [9]:
#transposing data - rows to columns to rows - wide format to long format

#import departments data
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

#check df_dep
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [10]:
#transpose df_dep from wide format to long format
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [11]:
#create new df_dep transposed 'df_dep_t'
df_dep_t = df_dep.T

#display new df_dep_t
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [12]:
#create df index column
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [13]:
#create new column headers
#create object that contains the first row of the dataframe
new_header = df_dep_t.iloc[0]

#check object/variable
new_header

0    department
Name: department_id, dtype: object

In [14]:
#remove the first row in the dataframe - as it is now the new header
df_dep_t_new = df_dep_t[1:]

#check df_dep_t_new
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [15]:
#Add the new header
df_dep_t_new.columns = new_header

#check df_dep_t_new
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#04. Data Dictionaries

In [16]:
#turn a dataframe into a dictionary
data_dict = df_dep_t_new.to_dict('index')

#check new dictionary variable
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [17]:
#check df_prods to identify first row department_id 
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [18]:
#identify department_id for row 0 using data_dict
print(data_dict.get('19'))

{'department': 'snacks'}


#05. Subsetting - creating a smaller data set from the whole set using a filter

In [19]:
#create filtered df (df_snacks) using department_id 19 in the df_prods dataframe
df_snacks =  df_prods[df_prods['department_id']==19]

#check df_snacks
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [22]:
#create filtered df (df_snacks_2) using loc function
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

#check df_snacks
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [23]:
#create filtered df (df_snacks_3) using isin function - allows for checking multiple id numbers [17,18,19]
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

#check df_snacks
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

#06. Task Questions

In [24]:
#changing df_ords user_id from numeric to string - not to be used in calculations as it is a key/label
df_ords['user_id'] = df_ords['user_id'].astype('str')

#check df_ords 'user_id' data type
df_ords['user_id'].dtype

dtype('O')

In [28]:
#renaming df_ords variable '
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

#check df_ords columns
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [29]:
#find frequencies of df_ord 'order_hour_of_day'
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#The busiest hour of the day for orders is 10:00 (10am) with 288,418 orders.

In [31]:
#identify department_id for value 4 using data_dict as defined previously
print(data_dict.get('4'))

{'department': 'produce'}


#The department_id data dictionary for 4 is produce.

In [32]:
#create filtered df (df_breakfast) using loc function
df_breakfast = df_prods.loc[df_prods['department_id'] == 14]

#check df_breakfast
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [33]:
#display number rows, columns for df_breakfast
df_breakfast.shape

(1116, 5)

In [34]:
#create multi filtered df (df_dinner_party) using isin function - alcohol[5], deli[20], beverages[7], meat/seafood[12]
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

#check df_dinner_party
df_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

In [36]:
#display number rows, columns for df_dinner_party
df_dinner_party.shape

(7650, 5)

#The last df created is df_dinner_party and has 7650 rows.

In [38]:
#create filtered df (df_user_1) for user_id 1
df_user_1 = df_ords[df_ords['user_id']== '1']

#check df_user_1
df_user_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_last_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0  
10                   14.0

In [39]:
#diplay descriptive analysis of df_user_1
df_user_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_last_order  
count              10.000000  
mean               19.000000  
std                 9.030811  
min                 0.000000  
25%                14.250000  
50%                19.500000  
75%                26.250000  
max                30.000000

In [40]:
#diplay number rows, columns for df_user_1
df_user_1.shape

(11, 6)

In [41]:
#export df_ords as "orders_wrangled.csv"
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [42]:
#export df_dep_t_new as "departments_wrangled.csv"
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))